### PR TITLE
Schemas are now only validated once across threads.

### LINKED ISSUE #1373: Validating Realm schema should only be done once
**Problem**:
1) Background thread opens Realm and begins a long living write transaction.
2) UI thread calls `Realm.getInstance()` for the first time.
3) Result: UI thread is locked until background thread completes because `initializeRealm` requires a write transaction.

**Reason**:
Our Realm cache is thread local and doesn't take into account that a Realm on another thread might have validated the schema already.

**Solution**:
We should cache the result of `mediator.validateTable(modelClass, realm.sharedGroup.getTransaction());`. Doing so will not only prevent unnecessary locking of the UI thread, but it will also speed up opening Realms considerable when doing so for the first time on new threads.

### PR DESCRIPTION
**Pull request by cmelchior**
Fixes #1373 

Prevents a long living write transaction on a background thread blocking opening Realm on the UI thread if it is the same file, as we then are sure that the schema has already been validated.

@realm/java

### CODE REVIEW

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 1]
```diff
@@ -27,6 +27,7 @@
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
+import java.sql.Time;
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: Are you sure you are importing the right class here?

File: realm/src/main/java/io/realm/Realm.java [hunk 2]
```diff
@@ -545,6 +551,7 @@ private static void initializeRealm(Realm realm) {
                 mediator.validateTable(modelClass, realm.transaction);
                 realm.columnIndices.addClass(modelClass, mediator.getColumnIndices(modelClass));
             }
+            validatedRealmFiles.put(realm.getPath(), true);
```
Comment (emanuelez) [inline realm/src/main/java/io/realm/Realm.java]: Use `Boolean.TRUE` so you don't have to rely on autoboxing

File: realm/src/main/java/io/realm/Realm.java [hunk 3]
```diff
@@ -128,6 +128,9 @@
         }
     };
 
+    // List of Realm files that has already been validated
+    private static final Map<String, Boolean> validatedRealmFiles = new ConcurrentHashMap<String, Boolean>();
```
Comment (emanuelez) [inline realm/src/main/java/io/realm/Realm.java]: How about using `Collections.newSetFromMap`? http://docs.oracle.com/javase/6/docs/api/java/util/Collections.html#newSetFromMap%28java.util.Map%29

**Comment by beeender** [comment]
Any good idea to handle the multi-process case?

**Comment by cmelchior** [comment]
This should not really effect multi-process (nor solve it). I would much rather take that discussion in #1300. This does however prevent the number of times we do require a write transaction which will also help with multi-process support.

**Comment by emanuelez** [comment]
:+1:

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 4]
```diff
@@ -1818,4 +1820,40 @@ public void testDeletingRealmAlsoClearsConfigurationCache() throws IOException {
         testRealm = Realm.getInstance(newConfig);
         assertNotNull(testRealm);
     }
+
+    public void testValidateSchemasOverThreads() throws InterruptedException, TimeoutException, ExecutionException {
```
Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: Please add a oneliner comment that explains what the test does.

File: realm/src/main/java/io/realm/Realm.java [hunk 5]
```diff
@@ -1477,6 +1487,7 @@ public static void migrateRealm(RealmConfiguration configuration, RealmMigration
             realm.beginTransaction();
             realm.setVersion(realmMigration.execute(realm, realm.getVersion()));
             realm.commitTransaction();
+            validatedRealmFiles.remove(configuration.getPath());
```
Comment (kneth) [inline realm/src/main/java/io/realm/Realm.java]: `validatedRealmFiles.add()` is done within a write transaction. Should `remove()` also be done within a write transaction?

Comment (cmelchior) [inline realm/src/main/java/io/realm/Realm.java]: While there is a race condition in close that could cause this to be called twice, that doesn't really matter as calling remove on something that isn't there would be a no-op. Also the backing set is Thread safe, so no there shouldn't be any need for a write transaction.

**Comment by cmelchior** [comment]
Updated with feedback from @nhachicha

**Comment by beeender** [comment]
:+1:

**Pull request merged**